### PR TITLE
Add D string overloads to methods in StringTable

### PR DESCRIPTION
**Pull request by jacob-carlborg**
This will help reduce the surface of where C strings and pointer length pairs are used.

### CODE REVIEW

File: src/dmd/root/stringtable.d [hunk 1]
```diff
@@ -111,6 +111,11 @@ public:
         return getValue(table[i].vptr);
     }
 
+    inout(StringValue)* lookup(const(char)[] str) inout nothrow pure
```
Comment (JinShil) [inline src/dmd/root/stringtable.d]: I hate to nitpick, but new public functions should have full DDoc.

Comment (jacob-carlborg) [reply src/dmd/root/stringtable.d]: Done.

File: src/dmd/root/stringtable.d [hunk 2]
```diff
@@ -103,6 +103,17 @@ public:
...
+     * Looks up the given string in the string table and returns its associated
+     * value.
+     *
+     * Params:
+     *  s = the string to look up
+     *  length = the length of $(D_PARAM s)
+     *
+     * Returns: the string's associated value, or `null` if the string doesn't
+     *  exist in the string table
+     */
```
Comment (wilzbach) [inline src/dmd/root/stringtable.d]: FWIW the DStyle we use for new symbols at Druntime/Phobos says:

> Documentation comments should not have leading stars on each line.

https://dlang.org/dstyle.html#phobos_documentation

**Review by wilzbach — APPROVED** [review]

File: src/dmd/root/stringtable.d (hunk 2 above)
Comment (jacob-carlborg) [reply src/dmd/root/stringtable.d]: 😞

Comment (jacob-carlborg) [reply src/dmd/root/stringtable.d]: It looks so weird. I have a feeling that most of the code out there is not using this style.

Comment (jacob-carlborg) [reply src/dmd/root/stringtable.d]: Fixed.

File: src/dmd/root/stringtable.d [hunk 3]
```diff
@@ -103,6 +103,17 @@ public:
         pools = null;
     }
 
+    /**
+    Looks up the given string in the string table and returns its associated
+    value.
+
+    Params:
+     s = the string to look up
+     length = the length of $(D_PARAM s)
```
Comment (wilzbach) [inline src/dmd/root/stringtable.d]: Technically `str` would need to be documented too (as it gets dittoed into this documentation.)

**Review by JinShil — APPROVED** [review]

File: src/dmd/root/stringtable.d (hunk 3 above)
Comment (jacob-carlborg) [reply src/dmd/root/stringtable.d]: We need a better way to inherit documentation that `ditto`. Fixed.

**Comment by jacob-carlborg** [comment]
Do these new methods need to be placed at after all C++ methods sine they cannot appear in the C++ header?

**Comment by jacob-carlborg** [comment]
@JinShil @wilzbach ping.

**Comment by JinShil** [comment]
> Do these new methods need to be placed at after all the C++ methods sine they cannot appear in the C++ header?

Yeah, that may be nice.  I won't hold up this PR on it though, and it can always be done as a followup.

**Comment by JinShil** [comment]
Semaphore CI seems stuck, so I'm closing and re-opening.

**Comment by jacob-carlborg** [comment]
>> Do these new methods need to be placed at after all the C++ methods sine they cannot appear in the C++ header?

> Yeah, that may be nice

Actually, since these functions are not virtual and don't occupy a slot in the vtable the order doesn't matter [1]. In that case I think it makes more sense to have the overloads next to each other.

[1] https://github.com/dlang/dmd/pull/8550#issuecomment-411880216

**Pull request merged**